server/devdocs/index.js

```diff
@@ -200,34 +253,73 @@ module.exports = function() {
 	// return the HTML content of a document (assumes that the document is in markdown format)
 	app.get( '/devdocs/service/content', ( request, response ) => {
 		let path = request.query.path;
+		const component = request.query.component;
 
-		if ( ! path ) {
+		if ( ! ( path || component ) ) {
 			response
 				.status( 400 )
-				.send( 'Need to provide a file path (e.g. path=client/devdocs/README.md)' );
+				.send( 'Need to provide a file path (e.g. path=client/devdocs/README.md) or component (e.g. component=author-selector)' );
 			return;
 		}
 
-		if ( ! /\.md$/.test( path ) ) {
-			path = fspath.join( path, 'README.md' );
-		}
+		// if we are getting readme contents, the old way ;)
+		if ( path ) {
+			if ( ! /\.md$/.test( path ) ) {
+				path = fspath.join( path, 'README.md' );
+			}
 
-		try {
-			path = fs.realpathSync( fspath.join( root, path ) );
-		} catch ( err ) {
-			path = null;
-		}
+			try {
+				path = fs.realpathSync( fspath.join( root, path ) );
+			} catch ( err ) {
+				path = null;
+			}
 
-		if ( ! path || path.substring( 0, root.length + 1 ) !== root + '/' ) {
-			response
-				.status( 404 )
-				.send( 'File does not exist' );
-			return;
-		}
+			if ( ! path || path.substring( 0, root.length + 1 ) !== root + '/' ) {
+				response
+					.status( 404 )
+					.send( 'File does not exist' );
+				return;
+			}
 
-		const fileContents = fs.readFileSync( path, { encoding: 'utf8' } );
+			const fileContents = fs.readFileSync( path, { encoding: 'utf8' } );
 
-		response.send( marked( fileContents ) );
+			response.send( marked( fileContents ) );
+		}
+
+		if ( component ) {
+			// we need to search the entire component tree for this guy ... :(
```

Comment:
I wonder whether we can avoid this by reusing/combining code from the [components usage stats feature](https://github.com/Automattic/wp-calypso/search?utf8=%E2%9C%93&q=components-usage-stats). This may help with react-docgen stuff, as well as potentially avoiding all these `fs` traversal methods.